nat/traversal/pinger.go

```diff
@@ -33,69 +36,92 @@ const prefix = "[NATPinger] "
 const pingInterval = 200
 const pingTimeout = 10000
 
-// ConfigParser is able to parse a config from given raw json
-type ConfigParser interface {
-	Parse(config json.RawMessage) (ip string, port int, err error)
-}
-
 // Pinger represents NAT pinger structure
 type Pinger struct {
-	localPort      int
-	pingTarget     chan json.RawMessage
-	pingReceived   chan struct{}
+	pingTarget     chan *Params
 	pingCancelled  chan struct{}
 	natEventWaiter NatEventWaiter
 	configParser   ConfigParser
-	stop           chan struct{}
 	once           sync.Once
+	natProxy       natProxy
+	portPool       portSupplier
+	consumerPort   int
 }
 
 // NatEventWaiter is responsible for waiting for nat events
 type NatEventWaiter interface {
 	WaitForEvent() Event
 }
 
+// ConfigParser is able to parse a config from given raw json
+type ConfigParser interface {
+	Parse(config json.RawMessage) (ip string, port int, serviceType services.ServiceType, err error)
+}
+
+type portSupplier interface {
+	Acquire() (port.Port, error)
+}
+
 // NewPingerFactory returns Pinger instance
-func NewPingerFactory(waiter NatEventWaiter, parser ConfigParser) *Pinger {
-	target := make(chan json.RawMessage)
-	received := make(chan struct{})
+func NewPingerFactory(waiter NatEventWaiter, parser ConfigParser, proxy natProxy, portPool portSupplier) *Pinger {
+	target := make(chan *Params)
 	cancel := make(chan struct{})
-	stop := make(chan struct{})
 	return &Pinger{
 		pingTarget:     target,
-		pingReceived:   received,
 		pingCancelled:  cancel,
 		natEventWaiter: waiter,
 		configParser:   parser,
-		stop:           stop,
+		natProxy:       proxy,
+		portPool:       portPool,
 	}
 }
 
-// Start starts NAT pinger and waits for pingTarget to ping
+type natProxy interface {
+	handOff(serviceType services.ServiceType, conn *net.UDPConn)
+	registerServicePort(serviceType services.ServiceType, port int)
+	isAvailable(serviceType services.ServiceType) bool
+}
+
+// Params contains session parameters needed to NAT ping remote peer
+type Params struct {
+	RequestConfig json.RawMessage
```

Comment:
Why is it called `RequesConfig`? 
Is it a config of the request?

Reply:
This config comes with session setup request.